core/src/main/java/com/linecorp/armeria/client/retry/HttpStatusBasedRetryStrategy.java

```diff
@@ -0,0 +1,93 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.client.retry;
+
+import java.util.concurrent.CompletableFuture;
+
+import org.reactivestreams.Subscriber;
+import org.reactivestreams.Subscription;
+
+import com.google.common.collect.Streams;
+
+import com.linecorp.armeria.common.HttpHeaders;
+import com.linecorp.armeria.common.HttpObject;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.HttpStatus;
+import com.linecorp.armeria.common.HttpStatusClass;
+
+/**
+ * Provides the {@link RetryStrategy} that decides to retry the request again using HTTP statuses.
+ */
+public class HttpStatusBasedRetryStrategy implements RetryStrategy<HttpRequest, HttpResponse> {
+
+    private final Iterable<HttpStatus> retryStatuses;
+
+    /**
+     * Create a new instance that decides to retry the request if the responded status matches one of the
+     * statuses in {@code statusList}.
+     *
+     * @param statusList the list of HTTP status to retry.
+     */
+    protected HttpStatusBasedRetryStrategy(Iterable<HttpStatus> statusList) {
+        this.retryStatuses = statusList;
+    }
+
+    @Override
+    public CompletableFuture<Boolean> shouldRetry(HttpRequest request, HttpResponse response) {
+        final CompletableFuture<Boolean> future = new CompletableFuture<>();
+        response.subscribe(new Subscriber<HttpObject>() {
+            private Subscription subscription;
+            private boolean retry;
+
+            @Override
+            public void onSubscribe(Subscription s) {
+                subscription = s;
+                s.request(Long.MAX_VALUE);
+            }
+
+            @Override
+            public void onNext(HttpObject o) {
+                if (o instanceof HttpHeaders) {
+                    final HttpHeaders headers = (HttpHeaders) o;
+                    final HttpStatus status = headers.status();
+                    if (status != null && status.codeClass() != HttpStatusClass.INFORMATIONAL) {
+                        if (Streams.stream(retryStatuses).anyMatch(retryStatus -> {
+                            final HttpStatus resStatus = headers.status();
+                            return resStatus != null && resStatus.code() == retryStatus.code();
+                        })) {
+                            retry = true;
+                        }
+                        subscription.cancel();
+                        future.complete(retry);
+                    }
+                }
+            }
+
+            @Override
+            public void onError(Throwable t) {
+                future.complete(retry);
+            }
+
+            @Override
+            public void onComplete() {
+                future.complete(retry);
```

Comment:
How about using `closeFuture().whenComplete()` instead of `onError()` and `onComplete()`?